### DISCUSSION TITLE
DataflowLinkOptions.PropagateCompletion not working as expected

### DISCUSSION THREAD

**Discussion by rambodghx** [dotnet/runtime#91725] [General] [answered]
Hi everyone ,
meanwhile I trying to implement some pipelines with IDataFlow , I faced with situation that I'm not sure is that bug or what so I decided to open discussion to find out is there any issue.
`var ac = new TransformBlock<int, int>(i => i * 2);`
`var ac2 = new ActionBlock<int>(i =>{});`
`ac.LinkTo(ac2,new DataflowLinkOptions{PropagateCompletion = true});`
`ac.Post(10);`
`ac.Post(12);`
`ac2.Complete();`

The result is that the `ac2` will not propagate completion to  `ac` and `await ac.Completion` will block.
However the description of this property in docs is saying _"whether the linked target will have completion and faulting notification propagated to it automatically"_
Is there any misunderstanding about this property?

**Comment by svick** [comment, marked as answer]
`PropagateCompletion` works the other way around: you complete `ac` and then wait for `ac2` to complete:

```c#
var ac = new TransformBlock<int, int>(i => i * 2);
var ac2 = new ActionBlock<int>(i => { });
ac.LinkTo(ac2, new DataflowLinkOptions { PropagateCompletion = true });
ac.Post(10);
ac.Post(12);
ac.Complete();
await ac2.Completion;
```

**Reply by rambodghx** [reply]
I Got it ,
thanks.
Is there any way to propagate upward?

**Reply by svick** [reply]
Not automatically, but you could always wire that yourself.

Though I'm not sure that makes sense in the dataflow model: the source of the flow of data should know when it's done, not the target/sink.
